lib/PageScraper.js

```diff
@@ -0,0 +1,79 @@
+/* globals XPCOMUtils, Services, EventEmitter */
+"use strict";
+
+const simplePrefs = require("sdk/simple-prefs");
+const {PageMod} = require("sdk/page-mod");
+const {data} = require("sdk/self");
+const {parseHTMLText} = require("lib/MetadataParser");
+const {Cu} = require("chrome");
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyGetter(this, "EventEmitter", function() {
+  const {EventEmitter} = Cu.import("resource://devtools/shared/event-emitter.js", {});
+  return EventEmitter;
+});
+
+class PageScraper {
+  constructor() {
+    EventEmitter.decorate(this);
+    this.pageMod = null;
+    this.enabled = false;
+
+    simplePrefs.on(PageScraper.PAGE_SCRAPER_PREF, () => {
+      const newEnabledValue = simplePrefs.prefs[PageScraper.PAGE_SCRAPER_PREF];
+      if (newEnabledValue && !this.enabled) {
+        this.load();
+      } else if (!newEnabledValue && this.enabled) {
+        this.unload();
+      }
+    });
+    if (simplePrefs.prefs[PageScraper.PAGE_SCRAPER_PREF]) {
+      this.load();
+    }
+  }
+  parseAndSave(raw, url) {
+    parseHTMLText(raw, url)
+      .then(metadata => {
+        // Todo: save in store
+        console.log(metadata); // eslint-disable-line no-console
+        Services.obs.notifyObservers(null, PageScraper.PAGE_PARSED_NOTIF, JSON.stringify(metadata));
+      });
+  }
+  observe(subject, topic, data) {
+    this.emit(topic, data);
+  }
+  load() {
+    if (this.pageMod) {
+      return;
+    }
+    Services.obs.addObserver(this, PageScraper.PAGE_PARSED_NOTIF);
+    this.pageMod = PageMod({
```

Comment:
instead of using PageMod, you could take a look at the message manager's `loadFrameScript` command.
http://searchfox.org/mozilla-central/source/dom/base/nsIMessageManager.idl#416

As an experiment, it should be OK to use `PageMod` for this, however, as we get more performance-focused, it's worthwhile to look at what an implementation for Firefox would look like.

The PageMod infrastructure uses a set of abstractions (content loaders, workers, rules, etc) that are loaded for each tab created and aren't needed if we're using `include: "*"` anyway... the framescript needs to load regardless.

A framescript + message-manager based implementation should be fairly quick and easy, without the overhead.